tests/Basics/SupervisedBufferTest.cpp

```diff
@@ -0,0 +1,479 @@
+  }
+
+  {
+    ResourceUsageScope usageScope(monitor);
+    AqlValue smallValue;
+    std::size_t sizeBeforeLocal = 0;
+    std::size_t valueSizeLocal = 0;
+    {
+      Builder builder;
+      builder.openArray();
+      builder.add(Value(42));
+      builder.close();
+      ASSERT_EQ(monitor.current(), 0);
+
+      std::size_t preMonitor = monitor.current();
+      std::size_t sizeBefore = builder.size();
+      usageScope.increase(sizeBefore);
+      ASSERT_EQ(monitor.current(), preMonitor + sizeBefore);
+
+      sizeBeforeLocal = sizeBefore;
+      smallValue = AqlValue{builder.slice(), builder.size()};
+      valueSizeLocal = smallValue.memoryUsage();
+      monitor.increaseMemoryUsage(valueSizeLocal);
+      ASSERT_EQ(monitor.current(), sizeBefore + smallValue.memoryUsage());
+    }
+    ASSERT_EQ(monitor.current(), sizeBeforeLocal + valueSizeLocal);
+    monitor.decreaseMemoryUsage(valueSizeLocal);
+    smallValue.destroy();
+    ASSERT_EQ(monitor.current(), 0);
+  }
+}
+
+TEST(SupervisedbufferTest, ReuseSupervisedBufferAccountsMemory) {
+  auto& global = GlobalResourceMonitor::instance();
+  ResourceMonitor monitor{global};
+
+  AqlValue firstValue;
+  AqlValue secondValue;
+
+  {
+    SupervisedBuffer supervisedBuffer(monitor);
+    Builder builder(supervisedBuffer);
+    builder.openArray();
+    builder.add(Value(std::string(1024, 'a')));
+    builder.close();
+    firstValue = AqlValue{builder.slice(), builder.size()};
+
+    monitor.increaseMemoryUsage(firstValue.memoryUsage());
+    ASSERT_GE(monitor.current(), firstValue.memoryUsage());
+  }
+  ASSERT_EQ(monitor.current(), firstValue.memoryUsage());
+
+  {
+    SupervisedBuffer supervisedBuffer(monitor);
+    Builder builder(supervisedBuffer);
+    builder.openArray();
+    builder.add(Value(std::string(2048, 'b')));
+    builder.close();
+    secondValue = AqlValue{builder.slice(), builder.size()};
+
+    monitor.increaseMemoryUsage(secondValue.memoryUsage());
+    ASSERT_GE(monitor.current(),
+              firstValue.memoryUsage() + secondValue.memoryUsage());
+  }
+  ASSERT_EQ(monitor.current(),
+            firstValue.memoryUsage() + secondValue.memoryUsage());
+
+  // drop our manual accounting for both values before destroying them
+  monitor.decreaseMemoryUsage(firstValue.memoryUsage());
+  monitor.decreaseMemoryUsage(secondValue.memoryUsage());
+  firstValue.destroy();
+  secondValue.destroy();
+  ASSERT_EQ(monitor.current(), 0);
+}
+
+TEST(SupervisedbufferTest, SupervisedBuilderGrowthAndRecycle) {
+  auto& global = GlobalResourceMonitor::instance();
+  ResourceMonitor monitor{global};
+
+  {
+    ResourceUsageScope usageScope(monitor);
+    SupervisedBuffer supervisedBuffer(monitor);
+    Builder builder(supervisedBuffer);
+
+    // memory should be low >= size here
+    builder.openArray();
+    builder.add(Value(1));
+    builder.add(Value(2));
+    builder.add(Value(3));
+    builder.close();
+    AqlValue smallValue = AqlValue{builder.slice(), builder.size()};
```

Comment:
What is the benefit of using AqlValues here? They are not contributing to the accounting, only after manually adding them, but that is tested in the tests above at length.